juju/model.py

```diff
@@ -108,41 +108,50 @@ class Relation:
     def __init__(self, relation_name, relation_id, local_unit, backend, cache):
         self.relation_name = relation_name
         self.relation_id = relation_id
-        self._local_unit = local_unit
-        self._backend = backend
-        self._cache = cache
-        self._apps = None
-        self._units = None
-        self._data = None
+        self._apps = set()
+        self._units = []
+        for unit_name in backend.relation_list(relation_id):
+            unit = cache.get(Unit, unit_name)
+            self._units.append(unit)
+            self._apps.add(unit.app)
+        self.data = RelationData(relation_name, relation_id, local_unit, self.units, backend)
 
     @property
     def apps(self):
-        if self._apps is None:
-            self._apps = list({unit.app for unit in self.units})
-        return self._apps
+        return list(self._apps)
```

Comment:
Why are these returning a copy every time? Shouldn't we just have self.data and self.units be the real things?